src/mlir/consumer.h

```diff
@@ -0,0 +1,149 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "generator.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "parser/ast.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+#include <string>
+
+namespace mlir::verona
+{
+  /**
+   * Field offset maps between field names, types and their relative position.
+   *
+   * FIXME: This will need to be better encapsulated, but we'll expand it when
+   * we start implementing classes fully, and dynamic dispatch.
+   */
+  struct FieldOffset
+  {
+    llvm::SmallVector<llvm::StringRef, 1> fields;
+    llvm::SmallVector<Type, 1> types;
+  };
+
+  /**
+   * AST Consumer
+   */
+  class ASTConsumer
+  {
+    ASTConsumer(MLIRContext* context) : gen(context) {}
+
+    /// MLIR Generator
+    MLIRGenerator gen;
+
+    /// Map for each type which fields does it have.
+    std::map<llvm::StringRef, FieldOffset> classFields;
+
+    /// Function scope, for mangling names. 3 because there will always be the
+    /// root module, the current module and a class, at the very least.
+    llvm::SmallVector<llvm::StringRef, 3> functionScope;
+
+    /// HACK: This tracks assignment types for `select` functions without a
+    /// return type.
+    /// FIXME: Either `select` should have a type or we should track this in a
+    /// context variable of sorts.
+    Type assignTypeFromSelect;
+
+    /// AST aliases
+    using Ast = ::verona::parser::Ast;
+    using AstPath = ::verona::parser::AstPath;
+
+    // ===================================================== Helpers
+    // Methods for symbols, location and other helpers for building
+    // MLIR nodes.
+
+    /// Get builder from generator.
+    OpBuilder& builder()
+    {
+      return gen.getBuilder();
+    }
+
+    /// Get symbol table from generator.
+    SymbolTableT& symbolTable()
+    {
+      return gen.getSymbolTable();
+    }
+
+    /// Get node as a shared pointer of a sub-type
+    template<class T>
+    ::verona::parser::Node<T> nodeAs(::verona::parser::Ast from)
+    {
+      return std::make_shared<T>(from->as<T>());
+    }
+
+    /// Get location of an ast node.
+    Location getLocation(Ast ast);
+
+    /// Mangle function names. If scope is not passed, use functionScope.
+    std::string mangleName(
+      llvm::StringRef name, llvm::ArrayRef<llvm::StringRef> scope = {});
```

Comment:
```suggestion
    std::string mangleName(
      llvm::StringRef name, llvm::ArrayRef<llvm::StringRef> scope);
    std::string mangleName(
      llvm::StringRef name)
    {
      return mangleName(name, functionScope);
    }
```

(Avoids the need to check for the empty thing in the callee)